using_restricted_array.rb

```diff
@@ -6,44 +6,115 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
+
 def length(array)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] == nil
+    index += 1
+  end
+  return index
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] == nil
+    print "#{array[index]} "
+    index += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] > length
+    if array[index] == value_to_find
+      return true
+    else
+      return false
```

Comment:
There's a bug here.
If you return false here, line 39 will never get executed. In fact, the method will return false if the first element in the array is not the value_to_find. So, instead of returning false on line 38, you should do it after line 41, i.e. at the end of the loop. This is when you've looked at all elements in the array and none of them matched the value_to_find (and hence you didn't return true).